astra/src/main/java/com/slack/astra/graphApi/GraphBuilder.java

```diff
@@ -0,0 +1,150 @@
+    for (ZipkinSpanResponse span : spans) {
+      if (span.getId() == null) {
+        LOG.warn("Document={} cannot have missing id", span);
+        continue;
+      }
+
+      // Create a node from this span's metadata
+      Node childNode = createChildNodeFromSpan(span);
+      nodes.putIfAbsent(childNode.getId(), childNode);
+      spanIdToNodeId.put(span.getId(), childNode.getId());
+
+      // Track parent-child relationships for edge creation
+      String parentId = span.getParentId();
+      if (parentId != null) {
+        childrenByParentSpan
+            .computeIfAbsent(parentId, k -> new ArrayList<>())
+            .add(childNode.getId());
+      }
+    }
+
+    // Second pass: create edges from parent-child relationships
+    HashSet<Edge> edges = buildEdges(nodes, childrenByParentSpan, spanIdToNodeId);
+
+    return new Graph(new ArrayList<>(nodes.values()), edges);
+  }
+
+  /**
+   * Creates a Node from a Zipkin span using configured metadata extraction.
+   *
+   * <p>In default behavior, extracts service name from the remote endpoint on the span. Otherwise,
+   * uses the GraphConfig's tag mapping to populate node metadata from span tags.
+   *
+   * @param span The Zipkin span to convert to a node
+   * @return Node with metadata extracted from the span
+   */
+  private Node createChildNodeFromSpan(ZipkinSpanResponse span) {
+    Map<String, String> tags = span.getTags();
+    TreeMap<String, String> metadata = new TreeMap<String, String>();
+
+    if (this.defaultBehaviour) {
+      metadata.put("service", span.getRemoteEndpoint().getServiceName());
+    } else {
+      for (String key : this.config.getNodeMetadataTagMapping().keySet()) {
+        metadata.put(key, config.resolve(tags, key));
+      }
+    }
+
+    return new Node(metadata);
+  }
+
+  /**
+   * Builds edges representing parent-child relationships between nodes.
+   *
+   * <p>Creates edges by connecting parent spans to their child spans, establishing the dependency
+   * relationships in the service graph. Logs warnings for any missing parent or child nodes.
+   *
+   * @param nodes Map of node IDs to Node objects
+   * @param childrenByParentSpan Map of parent span IDs to lists of child node IDs
+   * @param spanIdToNodeId Map of span IDs to corresponding node IDs
+   * @return Set of edges representing service dependencies
+   */
+  private HashSet<Edge> buildEdges(
+      HashMap<String, Node> nodes,
+      HashMap<String, ArrayList<String>> childrenByParentSpan,
+      HashMap<String, String> spanIdToNodeId) {
+    HashSet<Edge> edges = new HashSet<>();
+
+    // Iterate through every parent
+    for (HashMap.Entry<String, ArrayList<String>> entry : childrenByParentSpan.entrySet()) {
+      String parentSpanId = entry.getKey();
+      // Get the node for the parent span
+      String parentNodeId = spanIdToNodeId.get(parentSpanId);
+      Node parentNode = parentNodeId != null ? nodes.get(parentNodeId) : null;
```

Comment:
you can just pass null to get, it will just return null, unless you've set null as a key.
```suggestion
      Node parentNode = nodes.get(parentNodeId);
```